File: frontend/app/api/upload/route.ts
import { NextResponse } from "next/server";
import { createClient } from "@supabase/supabase-js";
import { PDFLoader } from "@langchain/community/document_loaders/fs/pdf";
import { RecursiveCharacterTextSplitter } from "@langchain/textsplitters";
import { SupabaseVectorStore } from "@langchain/community/vectorstores/supabase";
// --- 1. Import Xenova (HuggingFace) ---
import { HuggingFaceInferenceEmbeddings } from "@langchain/community/embeddings/hf";
import { type Document } from "@langchain/core/documents";

export const maxDuration = 60;
export const runtime = "nodejs";

const SUPABASE_URL = process.env.NEXT_PUBLIC_SUPABASE_URL;
const SUPABASE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY;

if (!SUPABASE_URL || !SUPABASE_KEY) {
  throw new Error(
    "Missing Supabase env vars. Ensure SUPABASE_URL and SUPABASE_SERVICE_KEY are set."
  );
}

const supabase = createClient(SUPABASE_URL, SUPABASE_KEY);

export async function POST(req: Request) {
  try {
    const formData = await req.formData();
    const file = formData.get("file") as File | null;
    const userEmail = (formData.get("user_email") as string) || "test@example.com";

    if (!file) {
      return NextResponse.json({ error: "No file uploaded" }, { status: 400 });
    }

    console.log(`📄 Processing file: ${file.name}`);

    // --- parse pdf -> docs
    const loader = new PDFLoader(file);
    const docs = await loader.load();
    if (!docs || docs.length === 0) {
      return NextResponse.json({ error: "Could not parse text from PDF" }, { status: 400 });
    }

    // --- split
    const splitter = new RecursiveCharacterTextSplitter({
      chunkSize: 1000,
      chunkOverlap: 200,
    });
    const split_docs = await splitter.splitDocuments(docs);

    // --- attach metadata (⏱️ add timestamp)
    const uploadedAt = Date.now();
    const docsWithMetadata = split_docs.map((d: Document, i: number) => ({
      ...d,
      metadata: {
        ...d.metadata,
        user_id: userEmail,
        file_name: file.name,
        chunk_index: i,
        uploaded_at: uploadedAt, // ⏱️
      },
    }));

    console.log(`🧩 Created ${docsWithMetadata.length} chunks.`);

    // --- 2. Use Xenova Embeddings ---
    const embeddings = new HuggingFaceInferenceEmbeddings({
      apiKey: process.env.HF_API_KEY!, // create at https://huggingface.co/settings/tokens
      model: "sentence-transformers/all-MiniLM-L6-v2",
    });

    console.log("🪣 Attempting SupabaseVectorStore.fromDocuments...");
    const testEmbedding = await embeddings.embedQuery("test");
    console.log("🧠 Test embedding vector length:", testEmbedding.length);

    try {
      const vectorStore = await SupabaseVectorStore.fromDocuments(
        docsWithMetadata,
        embeddings,
        {
          client: supabase,
          tableName: "documents",
          queryName: "match_documents",
        }
      );

      console.log("✅ fromDocuments returned:", !!vectorStore);
      return NextResponse.json({
        message: "File processed and embeddings stored successfully",
        file_name: file.name,
        chunks_created: docsWithMetadata.length,
        uploaded_at: uploadedAt,
        expires_in: "120 seconds",
      });
    } catch (insertErr: any) {
      console.error("Error inserting via SupabaseVectorStore.fromDocuments:", insertErr);

      const details =
        insertErr && typeof insertErr === "object"
          ? {
              message: insertErr.message ?? String(insertErr),
              stack: insertErr.stack,
              response: (insertErr as any).response ?? undefined,
              status: (insertErr as any).status ?? undefined,
            }
          : String(insertErr);

      return NextResponse.json(
        {
          error: "Error inserting vectors into Supabase (detailed info in server logs).",
          insertErrorSummary: details,
          suggestion:
            "Common causes: network blocked, wrong Supabase URL/key, missing pgvector extension or missing 'match_documents' rpc, or insufficient DB permissions.",
        },
        { status: 502 }
      );
    }
  } catch (error: any) {
    console.error("❌ Upload error:", error);
    return NextResponse.json(
      { error: error.message || "Upload failed", stack: error.stack },
      { status: 500 }
    );
  }
}

File: frontend/api/upload/route.py
from fastapi import FastAPI, UploadFile, File, Form, HTTPException
from fastapi.responses import JSONResponse
import os
import shutil
import tempfile
from langchain_community.document_loaders import PyPDFLoader
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain_huggingface import HuggingFaceEmbeddings
from supabase.client import create_client, Client

app = FastAPI()

# --- Initialize Supabase and Embeddings ---
supabase_url = os.getenv("SUPABASE_URL")
supabase_key = os.getenv("SUPABASE_SERVICE_KEY")
if not supabase_url or not supabase_key:
    raise ValueError("Supabase URL/Key not set in environment.")

supabase_client: Client = create_client(supabase_url, supabase_key)

print("Loading local embedding model (all-MiniLM-L6-v2)...")
# This runs on Vercel's 1GB+ memory
embeddings = HuggingFaceEmbeddings(model_name="all-MiniLM-L6-v2")
print("Local embedding model loaded.")
# ---

# This function will be served at POST /api/upload
@app.post("/")
async def upload_file(
    user_email: str = Form(...),
    file: UploadFile = File(...)
):
    temp_dir = tempfile.gettempdir()
    temp_file_path = None
    try:
        temp_file_path = os.path.join(temp_dir, file.filename)
        with open(temp_file_path, "wb") as buffer:
            shutil.copyfileobj(file.file, buffer)

        loader = PyPDFLoader(temp_file_path)
        docs = loader.load()
        if not docs:
            return JSONResponse(status_code=400, content={"detail": "Could not load any content from the PDF."})

        text_splitter = RecursiveCharacterTextSplitter(chunk_size=1000, chunk_overlap=200)
        split_docs = text_splitter.split_documents(docs)

        contents = [doc.page_content for doc in split_docs]
        print(f"Embedding {len(contents)} chunks for user {user_email}...")
        doc_embeddings = embeddings.embed_documents(contents)

        data_to_insert = []
        for i, doc in enumerate(split_docs):
            data_to_insert.append({
                "user_id": user_email,
                "file_name": file.filename,
                "content": contents[i],
                "embedding": doc_embeddings[i],
                "metadata": doc.metadata
            })

        print(f"Storing {len(data_to_insert)} vectors in Supabase...")
        
        batch_size = 50
        for i in range(0, len(data_to_insert), batch_size):
            batch = data_to_insert[i:i + batch_size]
            response = supabase_client.table("documents").insert(batch).execute()
            if response.data is None:
                 raise Exception(f"Insert failed: {response.error.message if response.error else 'Unknown error'}")

        print(f"Successfully embedded and stored file {file.filename}.")
        
        return JSONResponse(
            status_code=200, 
            # Send back the original temp path as the "file_path"
            content={"file_path": temp_file_path, "message": "File processed and ready for questions."}
        )

    except Exception as e:
        print(f"Error during file processing: {e}")
        return JSONResponse(status_code=500, content={"detail": f"There was an error processing the file: {str(e)}"})
    finally:
        if temp_file_path and os.path.exists(temp_file_path):
            os.remove(temp_file_path)